entity-service/src/main/kotlin/com/egm/stellio/entity/service/EntityOperationService.kt

```diff
@@ -0,0 +1,139 @@
+import org.springframework.stereotype.Component
+import kotlin.streams.toList
+
+/**
+ * Service to work on a list of entities.
+ */
+@Component
+class EntityOperationService(
+    private val neo4jRepository: Neo4jRepository,
+    private val entityRepository: EntityRepository,
+    private val entityService: EntityService,
+    private val entitiesEntitiesGraphBuilder: EntitiesGraphBuilder
+) {
+
+    /**
+     * Splits [entities] by their existence in the DB.
+     */
+    fun splitEntitiesByExistence(entities: List<ExpandedEntity>): Pair<List<ExpandedEntity>, List<ExpandedEntity>> {
+        val existingEntitiesIds =
+            neo4jRepository.filterExistingEntitiesIds(entities.map { it.id })
+        return entities.partition {
+            existingEntitiesIds.contains(it.id)
+        }
+    }
+
+    /**
+     * Creates a batch of [entities].
+     *
+     * @return a [BatchOperationResult]
+     */
+    fun create(entities: List<ExpandedEntity>): BatchOperationResult {
+        val (graph, invalidRelationsErrors) = entitiesEntitiesGraphBuilder.build(entities)
+
+        val (naiveBatchResult, entitiesWithCircularDependencies) = createEntitiesWithoutCircularDependencies(graph)
+        val (circularCreateSuccess, circularCreateErrors) = createEntitiesWithCircularDependencies(
+            entitiesWithCircularDependencies.toList()
+        )
+
+        val success = naiveBatchResult.success.plus(circularCreateSuccess)
+        val errors = invalidRelationsErrors.plus(naiveBatchResult.errors).plus(circularCreateErrors)
+
+        return BatchOperationResult(ArrayList(success), ArrayList(errors))
+    }
+
+    private fun createEntitiesWithoutCircularDependencies(graph: Graph<ExpandedEntity, DefaultEdge>): Pair<BatchOperationResult, Set<ExpandedEntity>> {
+        val batchOperationResult = BatchOperationResult(arrayListOf(), arrayListOf())
+        val temporaryGraph = DirectedPseudograph<ExpandedEntity, DefaultEdge>(DefaultEdge::class.java)
+        Graphs.addGraph(temporaryGraph, graph)
+
+        /**
+         * Gets a list of leaves (i.e. with 0 successors).
+         */
+        fun <T, E> Graph<T, E>.getLeaves(): List<T> = vertexSet().filter { outDegreeOf(it) == 0 }
+
+        do {
+            val leaves = temporaryGraph.getLeaves()
+
+            val res = leaves.parallelStream().map {
+                try {
+                    entityService.createEntity(it)
+                    Pair(it.id, null)
+                } catch (e: BadRequestDataException) {
+                    Pair(null, BatchEntityError(it.id, arrayListOf(e.message)))
+                }
+            }
+
+            res.toList().fold(batchOperationResult, { result, (entity, error) ->
+                entity?.let { result.success.add(it) }
+                error?.let { result.errors.add(it) }
+                result
+            })
+
+            leaves.forEach {
+                temporaryGraph.removeVertex(it)
+            }
+        } while (leaves.isNotEmpty())
```

Comment:
If I got it right, you'll have a sort of empty iteration for the last loop (`leaves` set to empty list in L72, but you'll only exit in the while at the end)?

(not a big deal, but it confused me)

Reply:
Indeed you got it right.